### PR TITLE
Fly, use rotating buffer over list to store last 5 positions.

### PR DESCRIPTION
**Pull request by anvilvapre**
Fly stores the last the position over the lats 5 ticks to determine the distance travelled and whether flight is stuck. 

- Replace position buffer `List` with an rotating buffer. To avoid removing the head element from a list array each tick. Avoiding a ArrayCopy/Move each tick.
- Avoid addition of one WVec/Pos to Zero WVev/Pos. Trivial. 
- Tickfacing, 50% change of one less subtraction. Trivial. 
- `RotatingBuffer` can be reused in new contrail implementation. Could potentially also be used in input tap detection. (only used function tested)

Future possible improvements: `Map.DistanceAboveTerrain` is looked up more often by all flight activites/functions - during the same tick. For rectangular maps the operation is not that expensive. Functions already take many arguments - did not want to add it.

### CODE REVIEW

File: OpenRA.Game/Primitives/RotatingBuffer.cs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+
+namespace OpenRA.Primitives
+{
+	/// <summary>Fixed size rorating buffer backed by an array.</summary>
+	public class RotatingBuffer<T> : ICollection<T>, IEnumerable<T>
```
Comment (teinarss) [inline OpenRA.Game/Primitives/RotatingBuffer.cs]: A more common name would be RingBuffer?

File: OpenRA.Game/Primitives/RotatingBuffer.cs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+			if (count == 0)
+				throw new ArgumentException("Empty buffer");
+
+			return values[(start + count - 1) % values.Length];
+		}
+
+		public IEnumerator<T> GetEnumerator()
+		{
+			var c = count;
+			var index = start;
```
Comment (pchote) [inline OpenRA.Game/Primitives/RotatingBuffer.cs]: Shouldn't this be 0?

File: OpenRA.Game/Primitives/RotatingBuffer.cs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+		public T Last()
+		{
+			if (count == 0)
+				throw new ArgumentException("Empty buffer");
+
+			return values[(start + count - 1) % values.Length];
+		}
+
+		public IEnumerator<T> GetEnumerator()
```
Comment (pchote) [inline OpenRA.Game/Primitives/RotatingBuffer.cs]: It looks like this won't behave correctly if the underlying collection is modified between enumerator yields. Other collections check for this and throw an InvalidOperationException("Collection was modified; enumeration operation may not execute").

File: OpenRA.Game/Primitives/RotatingBuffer.cs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+		{
+			var c = count;
```
Comment (pchote) [inline OpenRA.Game/Primitives/RotatingBuffer.cs]: Unnecessary local variable after implementing the exception.

File: OpenRA.Game/Primitives/RotatingBuffer.cs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+			var index = start;
+			while (index < c)
```
Comment (pchote) [inline OpenRA.Game/Primitives/RotatingBuffer.cs]: This would be simpler as a for loop.

File: OpenRA.Game/Primitives/RotatingBuffer.cs (hunk 2 above)
Comment (anvilvapre) [reply OpenRA.Game/Primitives/RotatingBuffer.cs]: No, it should be the start/head of the buffer.

Comment (pchote) [reply OpenRA.Game/Primitives/RotatingBuffer.cs]: This won't work with the `while (index < c)` and `(start + index++)` below.

File: OpenRA.Game/Primitives/RotatingBuffer.cs (hunk 3 above)
Comment (anvilvapre) [reply OpenRA.Game/Primitives/RotatingBuffer.cs]: It would still always return the last element. But yes, the user would not be aware.

The down side is the need to check if the start and size changed each access. But if it's a generic container, then it is a good thing to do.

File: OpenRA.Game/Primitives/RotatingBuffer.cs (hunk 2 above)
Comment (anvilvapre) [reply OpenRA.Game/Primitives/RotatingBuffer.cs]: ah yes. i now see the issue. index should start at 0.

**Comment by anvilvapre** [comment]
Requested changes addressed.

**Review by AspectInteractive2 — DISMISSED** [review]
This looks fine to me. I am not sure how much of a difference these changes will make but they are reasonable changes.

File: OpenRA.Game/Primitives/RingBuffer.cs [hunk 6]
```diff
@@ -0,0 +1,144 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2022 The OpenRA Developers (see AUTHORS)
```
Comment (abcdefg30) [inline OpenRA.Game/Primitives/RingBuffer.cs]: This needs to be adjusted for #20595.

File: OpenRA.Game/Primitives/RingBuffer.cs [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+	/// <summary>Fixed size rorating buffer backed by an array.</summary>
+	public class RingBuffer<T> : ICollection<T>, IEnumerable<T>
+	{
+		readonly IComparer<T> comparer;
+		readonly T[] values;
+		int start;
+		int count;
+
+		public int Capacity => values.Length;
+		public int Count => count;
```
Comment (PunkPun) [inline OpenRA.Game/Primitives/RingBuffer.cs]:
```suggestion
		public int Count { get; private set; };
```

File: OpenRA.Mods.Common/Activities/Air/Fly.cs [hunk 8]
```diff
@@ -29,7 +28,7 @@ public class Fly : Activity
 		Target target;
 		Target lastVisibleTarget;
 		bool useLastVisibleTarget;
-		readonly List<WPos> positionBuffer = new List<WPos>();
+		readonly RingBuffer<WPos> history = new RingBuffer<WPos>(5);
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Activities/Air/Fly.cs]: IMO we should keep the name positionBuffer

File: OpenRA.Game/Primitives/RingBuffer.cs [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+namespace OpenRA.Primitives
+{
+	/// <summary>Fixed size rorating buffer backed by an array.</summary>
+	public class RingBuffer<T> : ICollection<T>, IEnumerable<T>
+	{
+		readonly IComparer<T> comparer;
+		readonly T[] values;
+		int start;
+		int count;
```
Comment (PunkPun) [inline OpenRA.Game/Primitives/RingBuffer.cs]: The names `start` and `count` are weird and not very descriptive. Could they be changed / or explained? Like Start renamed to `CursorPosition` or smth like that?

**Review by PunkPun — COMMENTED** [review]
I think buffer related changes should be separated into its own commit. This also needs a rebased as it conflicts with the current codestyle (I've suggested the fix).

File: OpenRA.Game/Primitives/RingBuffer.cs (hunk 9 above)
Comment (anvilvapre) [reply OpenRA.Game/Primitives/RingBuffer.cs]: ) they are the start index of the first element, and the number of elements in the buffer.
head is worse. cursor position is not what it is.

File: OpenRA.Mods.Common/Activities/Air/Fly.cs [hunk 10]
```diff
@@ -29,7 +28,7 @@ public class Fly : Activity
 		Target target;
 		Target lastVisibleTarget;
 		bool useLastVisibleTarget;
-		readonly List<WPos> positionBuffer = new();
+		readonly RingBuffer<WPos> history = new(5);
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Activities/Air/Fly.cs]:
```suggestion
		readonly RingBuffer<WPos> previousPositions = new(5);
```

File: OpenRA.Game/Primitives/RingBuffer.cs [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+		int start;
+
+		public int Capacity => values.Length;
+		public int Count { get; private set; }
+		public bool IsReadOnly => false;
+
+		public RingBuffer(int capacity, IComparer<T> comparer)
+		{
+			this.comparer = comparer;
+			#pragma warning disable IDE0090
```
Comment (abcdefg30) [inline OpenRA.Game/Primitives/RingBuffer.cs]: Why is this necessary?

File: OpenRA.Game/Primitives/RingBuffer.cs [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+			if (Count < values.Length)
+				Count++;
+			else
+				start = (start + 1) % values.Length;
+		}
+
+		public void Clear()
+		{
+			Array.Clear(values, 0, values.Length);
+			Count = 0;
```
Comment (abcdefg30) [inline OpenRA.Game/Primitives/RingBuffer.cs:52]: Doesn't this need to reset `start` as well?

File: OpenRA.Game/Primitives/RingBuffer.cs [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+		}
+
+		public bool Contains(T value)
+		{
+			var capacity = values.Length;
+			var end = start + Count;
+			for (var i = start; i < end; ++i)
+				if (comparer.Compare(values[i % capacity], value) == 0)
+					return true;
```
Comment (abcdefg30) [inline OpenRA.Game/Primitives/RingBuffer.cs:61]: Is it really necessary to start at `start` here? We don't care about the order.
```suggestion
			for (var i = 0; i < capacity; ++i)
				if (comparer.Compare(values[i], value) == 0)
					return true;
```

File: OpenRA.Game/Primitives/RingBuffer.cs [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+		public void CopyTo(T[] array, int arrayIndex)
+		{
+			if (array == null)
+				throw new ArgumentNullException(nameof(array));
+
+			if (arrayIndex < 0)
+				throw new ArgumentNullException(nameof(arrayIndex));
+
+			if (arrayIndex + Count >= array.Length)
+				throw new ArgumentException("Invalid array capactiy");
```
Comment (abcdefg30) [inline OpenRA.Game/Primitives/RingBuffer.cs]:
```suggestion
				throw new ArgumentException("Invalid array capacity");
```

File: OpenRA.Game/Primitives/RingBuffer.cs [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+
+			var di = arrayIndex;
```
Comment (abcdefg30) [inline OpenRA.Game/Primitives/RingBuffer.cs]:
```suggestion
			var destinationIndex = arrayIndex;
```

File: OpenRA.Game/Primitives/RingBuffer.cs [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+			var capacity = values.Length;
+			for (var i = start; i < end; ++i)
+				array[di++] = values[i % capacity];
+		}
+
+		public bool Remove(T value)
+		{
+			var capacity = values.Length;
+			var end = start + Count;
+			for (var i = start; i < end; ++i)
```
Comment (abcdefg30) [inline OpenRA.Game/Primitives/RingBuffer.cs:88]: Please add braces to this `for`.

File: OpenRA.Game/Primitives/RingBuffer.cs [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+		}
+
+		public T this[int pos]
+		{
+			get => values[(start + pos) % values.Length];
+
+			set
+			{
+				if (pos >= Count)
+					throw new ArgumentException($"Index out of bounds: {pos}");
```
Comment (abcdefg30) [inline OpenRA.Game/Primitives/RingBuffer.cs:111]: We don't check this for `get`.

File: OpenRA.Game/Primitives/RingBuffer.cs (hunk 11 above)
Comment (anvilvapre) [reply OpenRA.Game/Primitives/RingBuffer.cs]: couldn't find what the `new()` syntax for arrays. `new ()[size]` seemed wrong?

File: OpenRA.Game/Primitives/RingBuffer.cs (hunk 12 above)
Comment (anvilvapre) [reply OpenRA.Game/Primitives/RingBuffer.cs:52]: as the buffer can wrap.

Comment (abcdefg30) [reply OpenRA.Game/Primitives/RingBuffer.cs:52]: But if we reset the buffer, we can/should also reset the wrapping, right?

File: OpenRA.Game/Primitives/RingBuffer.cs (hunk 13 above)
Comment (anvilvapre) [reply OpenRA.Game/Primitives/RingBuffer.cs:61]: initially values are default. you can still compare them, but you do not know what the comparator will do. i.e. searching for CPos(0,0) would then return true when there are no entries yet.

Comment (anvilvapre) [reply OpenRA.Game/Primitives/RingBuffer.cs:62]: i should make Comparer a type argument. as this proved to be much faster in priority queue.

File: OpenRA.Game/Primitives/RingBuffer.cs (hunk 11 above)
Comment (abcdefg30) [reply OpenRA.Game/Primitives/RingBuffer.cs]: My IDE claims the suppression is not needed:
image
Seems like you already fixed the issue somewhen after adding the pragmas.

File: OpenRA.Game/Primitives/RingBuffer.cs (hunk 13 above)
Comment (abcdefg30) [reply OpenRA.Game/Primitives/RingBuffer.cs:61]: > initially values are default. you can still compare them, but you do not know what the comparator will do. i.e. searching for CPos(0,0) would then return true when there are no entries yet.

Ah yeah that makes sense.

File: OpenRA.Game/Primitives/RingBuffer.cs (hunk 11 above)
Comment (abcdefg30) [reply OpenRA.Game/Primitives/RingBuffer.cs]: I get no warning after removing the pragmas:
image

File: OpenRA.Game/Primitives/RingBuffer.cs (hunk 17 above)
Comment (anvilvapre) [reply OpenRA.Game/Primitives/RingBuffer.cs:111]: yeah, it seemed reasonable, to avoid the overhead.

File: OpenRA.Game/Primitives/RingBuffer.cs (hunk 12 above)
Comment (anvilvapre) [reply OpenRA.Game/Primitives/RingBuffer.cs:52]: it doesn't matter where the buffer starts.

**Comment by anvilvapre** [comment]
ping. fixups were done, apart from the bounds check in the getter, which will be unlikely and obviously tracable by the resulting exception.

**Comment by anvilvapre** [comment]
ping.

**Review by PunkPun — APPROVED** [review]

**Comment by PunkPun** [comment]
[Chnagelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/e65ec2f865a7ef2fea1454c98ac004f2b26f8c15...49e6f9a80f54203129758313e8c93110fd945f62)
